proposals/0000-Move-Platforms-to-Own-Packages.md

```diff
@@ -0,0 +1,273 @@
+---
+title: Move iOS and Android Specific Code to Their Own Packages
+author:
+- Ian Meikle
+date: today
+---
+
+# RFC0000: Move iOS and Android Specific Code to Their Own Packages
+
+## Summary
+
+Re-organize the React Native repo as a Yarn workspace monorepo, and move platform specific JS and native code for iOS and Android into their own packages.
+
+## Basic example
+
+Common APIs and components would remain in `react-native`:
+
+```js
+import {AppRegistry, Alert, View} from 'react-native';
+```
+
+Platform-specific components and APIs would be imported from the platform's module:
```

Comment:
What about a module that has 4 methods out of 5 that work on both platforms and wants to provide one, additional, e.g. `Alert.promptIOS`? Would the developer need to make an additional `Alert` module under `react-native-ios` package? Wouldn't that cause troubles for developers to decide what file to use? Should `Alert` from `react-native-ios` "extend" `Alert` from `react-native` to keep its methods? Just general questions

Reply:
I do think some kinds of revamp of `Alert` is very much needed. There's this `if` statement in `Alert` that basically requires a platform dev to make a platform fork of `Alert`. Example from RNWindows:

```js
    if (Platform.OS === 'ios') {
      if (typeof type !== 'undefined') {
        console.warn('Alert.alert() with a 5th "type" parameter is deprecated and will be removed. Use AlertIOS.prompt() instead.');
        AlertIOS.alert(title, message, buttons, type);
        return;
      }
      AlertIOS.alert(title, message, buttons);
    } else if (Platform.OS === 'android') {
      AlertAndroid.alert(title, message, buttons, options);
    } else if (Platform.OS === 'windows') {
      AlertWindows.alert(title, message, buttons, options);
    }
```

And because it depends on requiring `AlertIOS`, it looks like some kind of modification will need to be made to this file in the split.

Reply:
Doing a suffix on the method is still bad because just because Android doesn't support it doesn't mean that Windows/MacOS/Ubuntu/other third party platforms won't. Rather than baking it into the method name it should just be documented which platforms it supports.